File: windows/MoDi.App.Contracts/Connectivity/TransportContracts.cs
namespace MoDi.App.Contracts.Connectivity;

public enum TransportSessionState
{
    Idle,
    Discovering,
    Listening,
    Connecting,
    Connected,
    Closing,
    Closed,
    Failed,
}

public sealed record TransportDescriptor(
    TransportKind Kind,
    bool SupportsDiscovery,
    bool SupportsListening,
    bool SupportsConnecting,
    bool IsOptional)
{
    public static TransportDescriptor For(TransportKind kind) => kind switch
    {
        TransportKind.Lan => new(kind, true, true, true, false),
        TransportKind.WifiDirect => new(kind, true, true, true, true),
        TransportKind.Bluetooth => new(kind, false, true, true, true),
        TransportKind.Usb => new(kind, false, true, true, true),
        _ => throw new ArgumentOutOfRangeException(nameof(kind)),
    };
}

public readonly record struct TransportEndpoint
{
    private TransportEndpoint(string value) => Value = value;
    public string Value { get; }

    public static TransportEndpoint Parse(string value)
    {
        if (string.IsNullOrWhiteSpace(value) || value != value.Trim() || value.Length > 512)
            throw new FormatException("Transport endpoint must be a non-empty opaque identifier.");
        return new TransportEndpoint(value);
    }

    public override string ToString() => Value ?? string.Empty;
}

public sealed record TransportCandidate(
    TransportEndpoint Endpoint,
    string? DisplayName,
    PeerId? ConfirmedPeerId = null);

public interface ITransportSession : IAsyncDisposable
{
    TransportDescriptor Descriptor { get; }
    TransportSessionState State { get; }
    event Action<ReadOnlyMemory<byte>>? BytesReceived;
    Task<IReadOnlyList<TransportCandidate>> DiscoverAsync(CancellationToken cancellationToken);
    Task ListenAsync(CancellationToken cancellationToken);
    Task ConnectAsync(TransportEndpoint endpoint, CancellationToken cancellationToken);
    Task SendAsync(ReadOnlyMemory<byte> payload, CancellationToken cancellationToken);
    Task CloseAsync(CancellationToken cancellationToken);
}

public sealed class TransportRegistry
{
    private readonly Dictionary<TransportKind, Func<ITransportSession>> _factories = [];

    public IReadOnlyCollection<TransportKind> RegisteredKinds => _factories.Keys;

    public void Register(TransportKind kind, Func<ITransportSession> factory)
    {
        ArgumentNullException.ThrowIfNull(factory);
        if (!_factories.TryAdd(kind, factory))
            throw new InvalidOperationException($"Transport {kind} is already registered.");
    }

    public bool TryCreate(TransportKind kind, out ITransportSession? session)
    {
        if (!_factories.TryGetValue(kind, out var factory))
        {
            session = null;
            return false;
        }

        session = factory();
        return true;
    }
}

File: android/app/src/main/java/com/modi/connect/core/connectivity/TransportContracts.kt
package com.modi.connect.core.connectivity

enum class TransportSessionState {
    Idle,
    Discovering,
    Listening,
    Connecting,
    Connected,
    Closing,
    Closed,
    Failed,
}

data class TransportDescriptor(
    val kind: TransportKind,
    val supportsDiscovery: Boolean,
    val supportsListening: Boolean,
    val supportsConnecting: Boolean,
    val isOptional: Boolean,
) {
    companion object {
        fun forKind(kind: TransportKind): TransportDescriptor = when (kind) {
            TransportKind.Lan -> TransportDescriptor(kind, true, true, true, false)
            TransportKind.WifiDirect -> TransportDescriptor(kind, true, true, true, true)
            TransportKind.Bluetooth -> TransportDescriptor(kind, false, true, true, true)
            TransportKind.Usb -> TransportDescriptor(kind, false, true, true, true)
        }
    }
}

@JvmInline
value class TransportEndpoint private constructor(val value: String) {
    override fun toString(): String = value

    companion object {
        fun parse(value: String): TransportEndpoint {
            require(value.isNotBlank() && value == value.trim() && value.length <= 512) {
                "Transport endpoint must be a non-empty opaque identifier"
            }
            return TransportEndpoint(value)
        }
    }
}

data class TransportCandidate(
    val endpoint: TransportEndpoint,
    val displayName: String?,
    val confirmedPeerId: PeerId? = null,
)

interface TransportSession {
    val descriptor: TransportDescriptor
    val state: TransportSessionState
    var onBytesReceived: ((ByteArray) -> Unit)?
    suspend fun discover(): List<TransportCandidate>
    suspend fun listen()
    suspend fun connect(endpoint: TransportEndpoint)
    suspend fun send(payload: ByteArray)
    suspend fun close()
}

class TransportUnavailableException(kind: TransportKind) :
    IllegalStateException("Transport $kind is not registered")

class TransportRegistry {
    private val factories = linkedMapOf<TransportKind, () -> TransportSession>()
    val registeredKinds: Set<TransportKind> get() = factories.keys.toSet()

    fun register(kind: TransportKind, factory: () -> TransportSession) {
        check(kind !in factories) { "Transport $kind is already registered" }
        factories[kind] = factory
    }

    fun tryCreate(kind: TransportKind): Result<TransportSession> = runCatching {
        val factory = factories[kind] ?: throw TransportUnavailableException(kind)
        factory()
    }
}
